debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlValueConverters.java

```diff
@@ -187,6 +192,11 @@ public SchemaBuilder schemaBuilder(Column column) {
                     return Decimal.builder(0);
             }
         }
+        switch (column.jdbcType()) { // see https://issues.jboss.org/browse/DBZ-342
+            case Types.TIME:
+                if (adaptiveTimePrecision && column.length() <= 0) return Time.builder();
```

Comment:
@gunnarmorling that's true, but I tried to find a common implementation to fulfill the handling of Durations and the base resolution of java.time.Duration is nanosecs. I can change that. But microseconds for 839 h force to use Long / INT64 anyway, so I took nanoseconds as that fits in Long/INT64 too.
I can update the PR, divide values by 1000 and use MicroTime as builder, if you want to go with microsec presicision.

Reply:
> I can update the PR, divide values by 1000 and use MicroTime as builder, if you want to go with microsec presicision.

Yes, that'd be my preference. We should strive to make as less behavioral changes as possible which may affect (and break) existing consumers. Transmitting nano-seconds instead of micro-seconds is such change we should avoid.

In terms of the "connect" time precision mode being obsolete, that's also my understanding. IIUC, it's there essentially to keep compatability with consumers existing before 0.2 which introduced the "adaptive" mode to avoid precision loss for `TIME(N>3)` columns. So the same motivation as above. We surely can get rid of it at some point, likely 1.0 (whenever we feel we're ready for that).

The remaining question is should we do the change to `INT64` unconditionally (again potentially breaking clients expecting `INT32` only), or should we provide an option to keep `INT32` (which of course only should be used if values don't exceed that value range). I'd be curious about @rhauch's take on that.

About that type discussion, I agree it's something to consider, but let's better have that one separate from the change at hand.

Reply:
@gunnarmorling @rhauch what about introducing a third time precision mode? ("connect", "adaptive" as default and making the PR compatible, and the one I suggest?) 
I think on the sink/consumer side, when not doing Avro and not knowing if the INT32/64 means seconds, micro-/milli- or nanoseconds, is really confusing. 

We would really prefer an intuitive mapping INT32 --> always seconds, INT64 --> always nanos (or millis)

What do you think?

Reply:
I don't think it's really appropriate to use the Schema version, since that will almost certainly conflict with Avro schema evolution versions (if they're used).

Changing from `INT64` to `INT32` may cause schema incompatibility issues for people that are relying upon certain evolution compatibility constraints. I don't recall how Avro would view such a change -- is it backwards or forwards compatible?

Is this really another distinct time precision mode, or is it really more overriding the mapped Schema type for a given source type or (fully-qualified) column name?

Reply:
@gunnarmorling you mean that one?

@rhauch I would introduce a new time mode here, as this is different behaviour. Users may not realize that this is a breaking change in the interpretation of the value and values in the stream (or in Stream Processors or sinked data in a destination database) could be misinterpreted. But I will again double check this to be really sure we don't introduce more configuration complexity, if the information in the streams is not changed. Will double check the information in the stream to be 100% sure.

Reply:
> @gunnarmorling you mean that one?

Agh, yes :) I was looking for it three times, but somehow missed it...

Reply:
I agree there should be a new time/precision mode, allowing people to configure whether either to allow for the full range of values by conveying TIME columns as `int64` or to keep the current behavior, if needed.

> I think on the sink/consumer side, when not doing Avro and not knowing if the INT32/64 means seconds, micro-/milli- or nanoseconds, is really confusing.
> 
> We would really prefer an intuitive mapping INT32 --> always seconds, INT64 --> always nanos (or millis)

Ah, I'm beginning to understand where you're coming from. IMHO, a consumer shouldn't interpret the value solely on the Connect type (`int32` vs. `int64`), though. Instead, the semantical type name should be used, i.e. `io.debezium.time.Time`. The semantics of that type are that it always conveys numbers in milliseconds. And that's not going to change, no matter whether the "physical" type is `int32` or if we changed it to `int64`.

Of course this assumes the schema is present in the JSON. I don't think it's practical to have no schema info at all (how are consumers even going to know that a `int32`/`int64` column is representing a duration without any out-of-band information on the semantics of that column; for them it's just an arbitrary number).

N.b. already today, `int32` may represent days (`io.debezium.time.Date`)  or milli-seconds (`io.debezium.time.Date`). And `int64` may be milli-seconds or micro-seconds (see http://debezium.io/docs/connectors/mysql/#temporal-values). So you can't make any conclusion just from looking at the plain type without the semantical name associated to it.

But I see how it may be desirable to not change the actual type of `io.debezium.time.Time` depending on some setting. I still don't think a new type is needed, though, a we already have `io.debezium.time.MicroTime`.

So would this work:

* add a new value for `time.precision.mode`: `adaptive_compat` (better name welcome)
* the behavior for `adaptive` (the default) would change so that columns of `TIME(0-6)` will always be represented by `io.debezium.time.MicroTime` and `int64`
* the new value `adaptive_compat` would keep the current behavior (`TIME(0-3)` are represented by `io.debezium.time.Time` and `int32`)

Would that work for you, @rk3rn3r? If not, how about having a quick Google Hangout some time this week, it may be easier to converge discussing face to face :)

Reply:
@gunnarmorling can we try to avoid changing behavior of existing functionality, unless doing so will result in the new Avro schema changes that are *both* backward and forward compatible with the Avro schema generated from the previous version?

I'm fine with adding a new `time.precision.mode`, but *changing* the `adaptive` is likely to generate an Avro schema that is not backward or forward compatible with the Avro schema generated for the same table by the previous Debezium version.

Changing field types is part of the API of Debezium, so we need to be far more careful about not breaking existing deployments when they upgrade.

Reply:
@rhauch I agree on the need on being careful here, but I don't think it's that we break things light-heartedly here (hence my proposal to have an option for keeping the current behavior).

But IMHO the current `adaptive` mode is broken, as it choses a type (`int32`) which doesn't fit the entire range of potential values for `TIME(0-3)`. So we need to provide users with a mapping of that column type which fits all the potential values, i.e. `int64`. Now the question is which setting should be the default, going forward. Based on previous similar cases, I think the correct mapping should be the default behavior, i.e. `int64`. I don't mind too much whether of the following two we do:

* having a new precision mode mapping to `int64`, using that as default and keep `adaptive` for compatability reasons, carrying the old mapping
* having `adaptive` change its mapping to `int64`, keept it as the default and have a new mode for compatability reasons, carrying the old meaning

But either way, I think `int64` should be the default if nothing else configured. Yes, it'd alter the default behavior, but it really is required to be able to convey all the possible values of that column type. In addition to the default, there'll be the option for keeping the current behavior, e.g. when upgrading an existing connector and being sure that `int32` is enough.

This change is scheduled for 0.7 so this seems the best thing to do to me, of course with appropriate documentation in the change log.

In a way it's the same kind of change as the BIGINT UNSIGNED mapping ([DBZ-363](https://issues.jboss.org/projects/DBZ/issues/DBZ-363)), where we agreed on changing the default mapping to a more sensible approach in 0.7, with an option to keep the existing behavior if required.

Does this make sense? Or do you see a difference to this case here?

Reply:
It's fine, but I just wanted the decision to break compatibility to be explicit and intentional.

Reply:
Ok, seems we're on the same page then :) Yes, it is intentional and it will be explicit by having a prominent entry in the release notes (together with the BIGINT UNSIGNED change).